### DISCUSSION TITLE
Using Homebrew as the basis for a Linux distribution

### DISCUSSION THREAD

**Discussion by jepotter1-archive** [Homebrew/discussions#1624] [Linux] [answered]
Hello everyone!

I'm looking into making a Linux distribution and I was wondering if it might be possible to repurpose Homebrew as a package manager for it.

One of the first things that would need to be figured out is how to run on a system Ruby installation, and make the Cellar path `/usr/lib/Cellar` or something like that instead of `/home/linuxbrew/.linuxbrew`. Then it would be necessary to figure out how to change the formulae URL from `https://github.com/homebrew/linuxbrew-core` to a custom repository.

Is this technically feasible, and if so, how might it be done with the least divergence from upstream?

(btw if anyone is interested in this project you can join the Matrix channel: [#foxos:kde.org](https://matrix.to/#foxos:kde.org))

Thank you,
Jeremy

**Comment by gromgit** [comment, marked as answer]
Not a maintainer, but the answer to your question is a qualified "yes". Homebrew's not _designed_ to be a system package manager, but all package managers basically "put files in directories, and manage their availability", something which Homebrew has pretty much mastered.

Most of your concerns are easily addressed:
> how to run on a system Ruby installation

It currently just requires stock Ruby 2.6.3. As long as you can get it built and in your `PATH`, Homebrew will happily use it. The only things I can think of that may not work are `brew test`, `brew audit`, and everything else that depends on a Sorbet binary for your architecture. This is available (and automatically downloaded) for x86-64 Linux, but not for other Linux architectures.

> make the Cellar path `/usr/lib/Cellar` or something like that

You can actually [install it under `/usr` or any directory you wish](https://docs.brew.sh/Installation#untar-anywhere). Homebrew is smart enough to use the actual location of its installation during normal operations.

Of course, there will be a whole bunch of directories like `etc` that would normally be expected to be under the root directory. This can easily be addressed with `ln -s /usr/etc /etc`.

> figure out how to change the formulae URL from https://github.com/homebrew/linuxbrew-core to a custom repository

Those are also easily changed in source, but that's yet more divergence.

However, those are minor issues compared to these two:

**No `sudo` For You:** Homebrew as-is specifically [forbids running as root](https://docs.brew.sh/FAQ#why-does-homebrew-say-sudo-is-bad). This check is easily eliminated, but of course that results in a "divergence from upstream", and various parts of your system will have to run and be managed as root no matter what (e.g. the init process).

**Bottles:** Not all existing Homebrew bottles are relocatable to a non-standard directory prefix. If you insist on installing Homebrew anywhere else, you'll either force all your user to "use the source", or experience the joy of [rebuilding all the bottles and hosting them on your own tap](https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/) _and_ manually merging all future Homebrew core tap changes (since all `bottles` blocks would diverge by definition).

If you want to leverage Homebrew to the maximum extent possible, and without taking on a crap-ton of recurring work, you'll want to leave it installed under `/home/linuxbrew/.linuxbrew` as a user-level application package manager, possibly restricted to a special `homebrew` user. What that does to the rest of your system is something you'll have to work out yourself, but since you'll already need a Ruby runtime for Homebrew, your systemland is likely best written in some combo of bash/zsh, Ruby and C.

Good luck!

**Reply by sjackman** [reply]
https://github.com/Linuxbrew/docker/blob/master/scratch/Dockerfile

As a fun little proof-of-concept, I built a standalone Docker image for Homebrew with no base OS, where every installed file is installed by Homebrew for Linux, and the package manager is Homebrew. As Adrian mentioned, it's simplest to use `/home/linuxbrew/.linuxbrew` as the installation prefix, and then symlink `/home/linuxbrew/.linuxbrew/bin` to `/bin` and `/usr/bin`. Using Homebrew as the package manager for a new Linux distro makes perfect sense to me!

You'll need to package the Linux kernel, and other startup and base packages not currently package for Homebrew for Linux.

You'll need to make sure that `brew upgrade` works without breaking the whole system, when the package being upgrade suddenly disappears. It currently relies on the fact that critical, essential system packages are provided by the host OS and never break. Fixing that assumption would be useful for Homebrew on Linux running on older systems where the host packages are too old to be useful.

**Reply by sjackman** [reply]
Is the https://foxos.org domain live yet? It's not working for me.

**Reply by danielnachun** [reply]
This is another case where https://github.com/Homebrew/brew/issues/10846 could also make things easier.  Although we are still hashing it possible corner cases and would have to rebuild all non-relocatable bottles, in principle, this would actually allow you to use our bottles from any prefix.  That would make the whole thing even more flexible, although symlinking to the root file system to `/home/linuxbrew/.linuxbrew/` already should work well without this.

>You'll need to make sure that brew upgrade works without breaking the whole system, when the package being upgrade suddenly disappears. It currently relies on the fact that critical, essential system packages are provided by the host OS and never break. Fixing that assumption would be useful for Homebrew on Linux running on older systems where the host packages are too old to be useful.

I've occasionally seen this cause breakage even between Homebrew packages.  I think we could eventually use some sort of allowlist of formula where we use `brew link --overwrite` to link from the cellar instead of unlinking first.  I could be wrong about this, but I suspect that part of the reason we unlink a formula before installing is because when building from source, you could end up with some weird things happening if you don't.  But those concerns don't apply if only using bottles.

>You'll need to package the Linux kernel, and other startup and base packages not currently package for Homebrew for Linux.

Using Homebrew as the system package manager for a CLI-only VM wouldn't be that much of a stretch.  It would get significantly harder to make it a package manager for a bare-metal OS with a GUI, because you'd need to package not just the kernel, but also full desktop environments, including display managers, as well as splash screens, boot managers, etc.  All of this should be possible to build, but the main limitation is the difficulty in testing those packages in CI.  

We also don't currently have a repo for proprietary software on Linux (aside from scientific software - Brewsci does allow that).  That could be important for shipping some drivers, video codecs and popular proprietary software like Google Chrome and Dropbox.  

Although there's a lot of work that would be needed to realize this, I think the idea of Homebrew as a single package manager for all use cases in Linux from managing the whole system to just installing some software in a user's home folder on an HPC cluster is quite tantalizing.  One of most common frustrations I've seen new users encounter is the ecosystem fragmentation for installing software.  More than once I've been asked "Why can't I just use Homebrew to manage all my software?" and my answer is that one day that just might be possible!

**Reply by sjackman** [reply]
https://foxos.codeberg.page

> LLVM/Clang + musl libc toolchain

That's definitely not a supported target of Homebrew, and you / your users would need to rebuild everything from source. It's definitely not going to be pain-free. All the macOS bottles are built with Apple Clang, so that _should_ mostly work, maybe? Musl may be more trouble, depending on how compatible it is with Glibc.

**Reply by jepotter1-archive** [reply]
> Is the https://foxos.org domain live yet? It's not working for me.

@sjackman it's not live yet sorry. I'm working on acquiring the domain right now 😄

for now you can go to the direct link for the website: https://foxos.codeberg.page

**Reply by jepotter1-archive** [reply]
btw thank you all for your friendly responses! I really appreciate it 🥰

**Reply by danielnachun** [reply]
See here for a very in depth discussion I had with an LLVM dev about LLVM on Linux in our setup: https://github.com/Homebrew/linuxbrew-core/pull/23339#issuecomment-850214614.  The TLDR for this is that our LLVM toolchain cannot be made independent of GCC for several reasons:
1) Bugs and missing features in the LLVM build process that still pull in runtime dependences on the GCC toolchain.  
2) LLVM would need to be built as a 2-stage build (we may eventually implement this without trying to make it independent of GCC just to improve optimization of the LLVM binaries by building them with Clang).
3) It would break our ability to link the shared LLVM libraries which are not compiler runtimes (`libLLVM`, `liblldb`, etc.) to binaries built with GCC because we could end up linking to both `libc++` and `libstdc++`, which could create all sorts of problems, especially for something like symbol versioning.  

We could eventually ship an `llvm-standalone` formula which could be used to build binaries without GCC if we can fix the first 2 issues above, but we wouldn't be using that to build our own bottles, just providing it to users for use outside of Homebrew.  If a major Linux distro were to switch to pure LLVM builds we could look into supporting that in our build process, but that hasn't happened yet.

>Musl may be more trouble, depending on how compatible it is with Glibc.

I'd be very curious how `musl` handles symbol versioning.  All of our bottles currently require Glibc 2.23 or newer, and we install our own `glibc` if the host one is too old.  If `musl` is ABI-compatible with `glibc` and supports symbol versioning at or above 2.23, it will be more likely to work.  I'm not sure how often `glibc` regex is used in our binaries, but this could be a problem (https://wiki.musl-libc.org/functional-differences-from-glibc.html): 

>Finally, glibc’s regex uses a 32-bit regoff_t even on 64-bit archs. This is non-conforming to POSIX and precludes giving correct output for strings larger than 2GB. musl uses a correct type, but this renders the ABI of the regex functions incompatible on 64-bit archs. The ABI is compatible on 32-bit.  

I also noticed:

>no systemd, minimalistic BSD-style init

which would be an issue for a fair number of our bottles which expect to be able to link to the `systemd` libraries (though this can be disabled in a custom tap).

I'm curious to see how this project works out.  This was a very interesting conversation that prompted me to think about some issues I haven't previously considered in Homebrew.  Good luck with your efforts!
